lib/elixir/src/elixir_erl_pass.erl

```diff
@@ -599,6 +632,36 @@ translate_remote(Left, Right, Meta, Args, S) ->
       {{call, Ann, {remote, Ann, TLeft, TRight}, TArgs}, SC}
   end.
 
+merge_pairs(List1, List2) ->
+  IsDuplicatedKey = fun ({map_field_assoc, _, Key1, _}) ->
```

Comment:
Pairs could potentially contain other expressions, like `map_field_exact`. Here is a sample code that fails with the current implementation:

```elixir
defmodule Foo do
  defstruct [:a, :b]
end

foo = %Foo{a: 1, b: 2}

Map.merge(%{foo | a: 3}, %{b: 4})
```

Using the code in the current branch, it raises:

```elixir
** (FunctionClauseError) no function clause matching in anonymous fn/1 in :elixir_erl_pass.merge_pairs/2

    The following arguments were given to anonymous fn/1 in :elixir_erl_pass.merge_pairs/2:

        # 1
        {:map_field_exact, 5, {:atom, 0, :a}, {:integer, 0, 3}}

    (stdlib) lists.erl:1286: :lists."-filter/2-lc$^0/1-0-"/2
```

Reply:
Good catch. But ideally we would remove the merging altogether and let the Erlang compiler do it, as they already do it anyway, and we just take care of the obvious ones for warning purposes. It should be easier in Elixir because literals do not have annotations. :)

Reply:
I posted the comment and then I saw yours 😅 Moving the warning to Elixir-lang sounds good to me 👍